panoply/error_normalization.py

```diff
@@ -0,0 +1,98 @@
+from panoply import DataSource, PanoplyException
+from datetime import datetime
+
+
+ERROR_CODES_REGISTRY = {
+    400: 'Bad request',
+    401: 'Unauthorized',
+    403: 'Permissions error',
+    404: 'Resource is not found',
+    408: 'Timeout',
+    422: 'Parsing error',
+    429: 'Rate limit',
+    500: 'Internal exception',
+    501: 'Unclassified error',
+    506: 'SDK error in processing an error'
+}
+
+EXCEPTIONS_REGISTRY = {}
+
+
+class NormalizedException(PanoplyException):
+    def __init__(self, message, error_code, exception_cls,
+                 phase, source_type, source_id, database_id,
+                 retryable=True):
+        super().__init__(message, retryable)
+        self.message = message
+        self.error_code = error_code
+        self.phase = phase
+        self.source_type = source_type
+        self.source_id = source_id
+        self.database_id = database_id
+        self.exception_cls = exception_cls
+        self.created_at = datetime.utcnow()
+
+
+def set_internal_code(error_code: int) -> callable:
+    """ A decorator is used on exception classes to register mapping
+        of error and internal code from ERROR_CODES_REGISTRY.
+        Parameters
+        ----------
+        error_code : int
+           Error code from ERROR_CODES_REGISTRY.
+    """
+    def decorator(exception_cls: type) -> type:
+        if not issubclass(exception_cls, BaseException):
+            raise RuntimeError('Unable to set error code for non-Exception class')
+
+        EXCEPTIONS_REGISTRY[exception_cls] = error_code
+
+        return exception_cls
+
+    return decorator
+
+
+def handle_errors(phase: str) -> callable:
+    """ A decorator is used to normalize raised exceptions.
+       Parameters
+       ----------
+       phase : str
+           Equals to config / collect.
+    """
+    def _handle_errors(func: callable) -> callable:
+        def wrapper(*args, **kwargs) -> list:
+            try:
+                return func(*args, **kwargs)
+            except Exception as e:
+                # source object can be:
+                # 1. a first param in dynamic params methods (e.g. definition(source, options))
+                # 2. an attribute of the DataSource class (e.g. definition(self, params) -> source = self.source)
+                source_config = args[0]
+                if isinstance(source_config, DataSource):
+                    source_config = getattr(source_config, 'source', None)
+
+                error_code = EXCEPTIONS_REGISTRY.get(type(e))
+
+                if error_code is None:
+                    error_code = 501
+
+                if error_code not in ERROR_CODES_REGISTRY:
+                    error_code = 506
+
+                details = {
+                    'message': str(e),
+                    'error_code': error_code,
+                    'exception_cls': str(type(e)),
```

Comment:
I believe that this will leave you with something similar to this: `<class '__main__.CustomException'>`. If you'll use `e.__class__.__name__` you should get just the name. I think this works for both python 2 and 3.

Reply:
@yonivy I think that sometimes the module from which the error originated might be useful as well. So to make the value of `exception_cls` prettier I can replace it with `f'{a.__class__.__module__}.{a.__class__.__name__}'`.

Reply:
Ok, whatever will help. My point was just that we can make things easier to parse for us.